File: apps/desktop/src/lib/omnibox.ts
import type { Bookmark, HistoryEntry, Tab } from "./ipc";
import tldList from "../../src-tauri/src/omnibox/tlds.txt?raw";

/** Most rows the address bar offers; past this the list hides the page for little gain. */
export const SUGGESTION_LIMIT = 8;

/** Delay before history and bookmarks are asked, so a burst of keystrokes costs one query. */
export const SUGGESTION_DEBOUNCE_MS = 120;

export type Suggestion =
  | { kind: "open" | "search" | "suggest"; url: string; title: string; favicon: null }
  | { kind: "tab"; url: string; title: string; favicon: string | null; tabId: string }
  | { kind: "bookmark" | "history"; url: string; title: string; favicon: string | null };

/** Schemes the engine loads itself when they are typed out in full. */
const ADDRESS_SCHEMES = new Set(["http", "https", "file", "about", "data", "blob", "view-source", "dive"]);

/**
 * Schemes the engine speaks (`INTERNAL` in external_link.rs). Any other
 * scheme typed out is another app's link, passed on so the "open in another
 * app" question is asked.
 */
const ENGINE_SCHEMES = new Set(["http", "https", "about", "blob", "data", "file", "javascript", "devtools", "chrome", "chrome-error", "chrome-extension", "chrome-untrusted", "view-source", "ws", "wss", "dive"]);

/** The top-level domains a bare name must end in, from the list the backend reads too. */
const TLDS = new Set(
  tldList
    .split("\n")
    .filter((line) => !line.startsWith("#"))
    .flatMap((line) => line.split(/\s+/))
    .filter(Boolean),
);

const IPV4 = /^(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.){3}(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)$/;
const LABEL = /^(?:[a-z0-9_-]|[^\p{ASCII}])+$/iu;

/**
 * Whether the backend will treat what was typed as an address rather than a
 * search. Mirrors `omnibox::classify` on the Rust side, which the address
 * bar's submit path always goes through; both are tested against the same
 * examples (src-tauri/src/omnibox/vectors.json), so a rule changed in one
 * has to change in the other.
 */
export function looksLikeUrl(input: string): boolean {
  const text = input.trim();
  if (!text || text.startsWith("?")) return false;
  if (text.startsWith("/") || text === "~" || text.startsWith("~/")) return true;
  if (/^[a-z]:[\\/]/i.test(text) && typeof navigator !== "undefined" && /Windows/.test(navigator.userAgent)) return true;
  return withScheme(text) || bareHost(text);
}

/** An address typed with a scheme the engine loads, or another app's link. */
function withScheme(text: string): boolean {
  let url: URL;
  try {
    url = new URL(text);
  } catch {
    return false;
  }
  const scheme = url.protocol.slice(0, -1);
  if (ADDRESS_SCHEMES.has(scheme)) return true;
  const rest = text.slice(text.indexOf(":") + 1);
  // "localhost:3000" parses as a scheme and a path; it is a host and a port.
  const portLike = /^\d+(?:[/?#]|$)/.test(rest);
  return scheme.length > 1 && !portLike && rest !== "" && !rest.startsWith(":") && !/\s/.test(text) && !ENGINE_SCHEMES.has(scheme);
}

/** A host typed without a scheme: an IP, this machine, a name with a port, or a name ending in a known domain. */
function bareHost(text: string): boolean {
  if (/\s/.test(text)) return false;
  const end = text.search(/[/?#]/);
  const split = splitPort(end === -1 ? text : text.slice(0, end));
  if (!split) return false;
  const { host, port } = split;
  if (host.startsWith("[")) return host.endsWith("]") && URL.canParse(`http://${host}`);
  if (IPV4.test(host)) return true;
  const labels = host.replace(/\.$/, "").toLowerCase().split(".");
  if (labels.some((label) => !LABEL.test(label))) return false;
  const tld = labels[labels.length - 1]!;
  // A single word is a search unless it is this machine or names a port:
  // "myserver:8080" is somebody's intranet, "12:30" a time.
  if (labels.length === 1) return tld === "localhost" || (port !== null && /\D/.test(tld));
  return port !== null || tld.startsWith("xn--") || /[^\p{ASCII}]/u.test(tld) || TLDS.has(tld);
}

/** `host:port` split apart, with a port that is a real one or none at all. */
function splitPort(authority: string): { host: string; port: number | null } | null {
  let host = authority;
  let port: string | null = null;
  if (authority.startsWith("[")) {
    const end = authority.indexOf("]");
    if (end < 0) return null;
    host = authority.slice(0, end + 1);
    const rest = authority.slice(end + 1);
    if (rest !== "") {
      if (!rest.startsWith(":")) return null;
      port = rest.slice(1);
    }
  } else {
    const colon = authority.indexOf(":");
    if (colon >= 0) {
      host = authority.slice(0, colon);
      port = authority.slice(colon + 1);
    }
  }
  if (!host) return null;
  if (port === null) return { host, port: null };
  if (!/^\d+$/.test(port) || Number(port) > 65535) return null;
  return { host, port: Number(port) };
}

/**
 * Host of `url`, or the empty string for anything that is not one. Dive's
 * own pages have no host worth showing, so they keep their whole address
 * ("dive://capture", not "capture").
 */
export function hostOf(url: string): string {
  try {
    const parsed = new URL(url);
    return parsed.protocol === "http:" || parsed.protocol === "https:" ? parsed.host : url;
  } catch {
    return "";
  }
}

/**
 * Where a row leads, as the list shows it: host and path without scheme,
 * query or a trailing slash ("localhost:8771/form.html"), so two pages on
 * one site can be told apart. Dive's own pages keep their whole address.
 */
export function placeOf(url: string): string {
  try {
    const parsed = new URL(url);
    if (parsed.protocol !== "http:" && parsed.protocol !== "https:") return url;
    const path = parsed.pathname === "/" ? "" : parsed.pathname.replace(/\/$/, "");
    return parsed.host + path;
  } catch {
    return "";
  }
}

/**
 * A row's name: its title, or its address when the title is empty or the
 * blank document's own name, which older history recorded before the page
 * had a title of its own.
 */
export function titleOf(entry: { title: string; url: string }): string {
  return entry.title && entry.title !== "about:blank" ? entry.title : entry.url;
}

/**
 * Whether typing `needle` is the start of the site `url` lives on, the way a
 * person types "gith" meaning github.com. Only the host counts, with or
 * without "www.", so a match in a path or a title does not qualify.
 */
export function leadsTo(needle: string, url: string): boolean {
  const host = hostOf(url).toLowerCase();
  if (!host || host === url) return false;
  return host.startsWith(needle) || host.replace(/^www\./, "").startsWith(needle);
}

function matches(query: string, ...fields: (string | null | undefined)[]) {
  return fields.some((field) => field?.toLowerCase().includes(query));
}

/**
 * The rows under the address bar for `query`, in the order they are offered:
 * the literal open-or-search row, then open tabs, bookmarks and history. A URL
 * appears once, wherever it is first seen -- a page that is open is offered as
 * a tab rather than again from history.
 */
export function buildSuggestions(
  query: string,
  { tabs, bookmarks, history, suggestions = [] }: { tabs: readonly Tab[]; bookmarks: readonly Bookmark[]; history: readonly HistoryEntry[]; suggestions?: readonly string[] },
  limit = SUGGESTION_LIMIT,
): Suggestion[] {
  const trimmed = query.trim();
  if (!trimmed) return [];
  const needle = trimmed.toLowerCase();
  const rows: Suggestion[] = [{ kind: looksLikeUrl(trimmed) ? "open" : "search", url: trimmed, title: trimmed, favicon: null }];
  const seen = new Set<string>();
  const add = (row: Suggestion) => {
    if (rows.length >= limit || seen.has(row.url)) return;
    seen.add(row.url);
    rows.push(row);
  };
  for (const tab of tabs) {
    if (matches(needle, tab.title, tab.url)) add({ kind: "tab", url: tab.url, title: titleOf(tab), favicon: tab.favicon, tabId: tab.id });
  }
  for (const bookmark of bookmarks) {
    if (matches(needle, bookmark.title, bookmark.url)) add({ kind: "bookmark", url: bookmark.url, title: titleOf(bookmark), favicon: bookmark.favicon });
  }
  for (const entry of history) {
    if (matches(needle, entry.title, entry.url)) add({ kind: "history", url: entry.url, title: titleOf(entry), favicon: entry.favicon });
  }
  // The engine's completions come last: what Dive already knows about beats a
  // guess, and they are what fills the list when it knows nothing. They are
  // searches, not addresses, so they never take the first row from a site the
  // letters lead to.
  for (const phrase of suggestions) {
    if (rows.length >= limit) break;
    if (rows.some((row) => row.kind !== "suggest" && row.url.toLowerCase() === phrase.toLowerCase())) continue;
    if (phrase.toLowerCase() === needle) continue;
    rows.push({ kind: "suggest", url: phrase, title: phrase, favicon: null });
  }
  // A few letters that begin a known site lead there on Enter, not to a web
  // search for those letters: "exam" with example.com open goes to the tab.
  // An address typed out in full keeps the literal row first, so Enter loads
  // it afresh even when a tab already shows it.
  if (!looksLikeUrl(trimmed)) {
    const site = rows.findIndex((row, index) => index > 0 && leadsTo(needle, row.url));
    if (site > 0) rows.unshift(...rows.splice(site, 1));
  }
  return rows.slice(0, limit);
}

/** Next highlight after an arrow key, wrapping at both ends. */
export function stepHighlight(current: number, count: number, delta: 1 | -1): number {
  if (count === 0) return 0;
  return (current + delta + count) % count;
}

File: apps/desktop/src-tauri/src/omnibox.rs
//! What the address bar makes of what was typed: an address to open, or
//! words to search for.
//!
//! The chrome decides the same thing on its own to label the first
//! suggestion "Open" or "Search" (`looksLikeUrl` in `src/lib/omnibox.ts`), and
//! a label that disagrees with what Enter then does is worse than no label.
//! So both read one list of top-level domains (`omnibox/tlds.txt`) and are
//! tested against one set of examples (`omnibox/vectors.json`); a rule
//! changed here has to change there too, and the shared examples say so.

use std::net::{Ipv4Addr, Ipv6Addr};
use std::path::PathBuf;

/// Schemes the engine loads itself when they are typed out in full.
const ADDRESS_SCHEMES: &[&str] = &[
    "http",
    "https",
    "file",
    "about",
    "data",
    "blob",
    "view-source",
    crate::engine::INTERNAL_SCHEME,
];

/// Top-level domains a bare name must end in to be taken for a site.
const TLDS: &str = include_str!("omnibox/tlds.txt");

/// Endings reserved for home and office networks that no public registry
/// hands out: nothing there has a certificate, so they are asked for over
/// plain http like the ones `https_only::is_local` knows.
const PRIVATE_TLDS: &[&str] = &["lan", "home", "corp", "intranet"];

/// What was typed, understood.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Typed {
    /// Open this.
    Address(url::Url),
    /// Search for these words.
    Search(String),
}

/// Read what was typed into the address bar; `None` when there is nothing
/// to act on.
pub fn classify(input: &str) -> Option<Typed> {
    let text = input.trim();
    if text.is_empty() {
        return None;
    }
    // A leading question mark is how a person says "search, even though
    // this looks like an address", as in every other browser.
    if let Some(rest) = text.strip_prefix('?') {
        let rest = rest.trim();
        return (!rest.is_empty()).then(|| Typed::Search(rest.to_owned()));
    }
    if let Some(url) = file_path(text)
        .or_else(|| with_scheme(text))
        .or_else(|| bare_host(text))
    {
        return Some(Typed::Address(url));
    }
    Some(Typed::Search(text.to_owned()))
}

/// A path on this machine: absolute, from the home folder, or (on Windows)
/// from a drive letter.
fn file_path(text: &str) -> Option<url::Url> {
    if text.starts_with('/') {
        return url::Url::from_file_path(text).ok();
    }
    if text == "~" || text.starts_with("~/") {
        let home = PathBuf::from(std::env::var_os("HOME")?);
        let path = home.join(text[1..].trim_start_matches('/'));
        return url::Url::from_file_path(path).ok();
    }
    let bytes = text.as_bytes();
    if cfg!(windows)
        && bytes.len() > 2
        && bytes[0].is_ascii_alphabetic()
        && bytes[1] == b':'
        && matches!(bytes[2], b'\\' | b'/')
    {
        return url::Url::from_file_path(text).ok();
    }
    None
}

/// An address typed with its scheme: one the engine loads, or another app's
/// (`mailto:`, `vscode:`, `zoommtg:`), which is passed on untouched so the
/// navigation reaches the "open in another app" question instead of a search.
fn with_scheme(text: &str) -> Option<url::Url> {
    let url = url::Url::parse(text).ok()?;
    let scheme = url.scheme();
    if ADDRESS_SCHEMES.contains(&scheme) {
        return Some(url);
    }
    let (_, rest) = text.split_once(':')?;
    // "localhost:3000" and "example.com:8080/x" parse as a scheme and a
    // path; they are a host and a port.
    let port_like = rest.starts_with(|c: char| c.is_ascii_digit())
        && rest
            .trim_start_matches(|c: char| c.is_ascii_digit())
            .chars()
            .next()
            .is_none_or(|c| matches!(c, '/' | '?' | '#'));
    let external = scheme.len() > 1
        && !port_like
        && !rest.is_empty()
        // "std::io::Read" is a question about Rust, not a link to an app.
        && !rest.starts_with(':')
        && !text.chars().any(char::is_whitespace)
        && crate::external_link::is_external(&url);
    external.then_some(url)
}

/// A host typed without a scheme (`example.com/docs`, `localhost:5173`,
/// `[::1]:8080`), given http when it is on this machine or the local network
/// and https everywhere else.
fn bare_host(text: &str) -> Option<url::Url> {
    if text.chars().any(char::is_whitespace) {
        return None;
    }
    let authority = &text[..text.find(['/', '?', '#']).unwrap_or(text.len())];
    let (host, port) = split_port(authority)?;
    let local = if let Some(v6) = host.strip_prefix('[') {
        let v6 = v6.strip_suffix(']')?;
        v6.parse::<Ipv6Addr>().ok()?;
        crate::https_only::is_local(v6)
    } else if let Ok(v4) = host.parse::<Ipv4Addr>() {
        v4.is_unspecified() || crate::https_only::is_local(host)
    } else {
        let name = host.strip_suffix('.').unwrap_or(host).to_ascii_lowercase();
        let labels: Vec<&str> = name.split('.').collect();
        if labels.iter().any(|label| !valid_label(label)) {
            return None;
        }
        let tld = labels.last().copied().unwrap_or_default();
        if labels.len() == 1 {
            // A single word is a search, unless it is this machine or names
            // a port: "myserver:8080" is somebody's intranet, "12:30" a time.
            let named =
                tld == "localhost" || (port.is_some() && tld.chars().any(|c| !c.is_ascii_digit()));
            if !named {
                return None;
            }
            true
        } else {
            // A port says "address" plainly enough; otherwise the ending has
            // to be a real one, or "node.js" and "3.14" would be opened.
            if port.is_none() && !known_tld(tld) {
                return None;
            }
            PRIVATE_TLDS.contains(&tld) || crate::https_only::is_local(&name)
        }
    };
    let scheme = if local { "http" } else { "https" };
    url::Url::parse(&format!("{scheme}://{text}")).ok()
}

/// `host:port` split apart, with a port that is a real one or none at all.
fn split_port(authority: &str) -> Option<(&str, Option<u16>)> {
    let (host, port) = if authority.starts_with('[') {
        let end = authority.find(']')?;
        let (host, rest) = authority.split_at(end + 1);
        match rest {
            "" => (host, None),
            _ => (host, Some(rest.strip_prefix(':')?)),
        }
    } else {
        match authority.split_once(':') {
            Some((host, port)) => (host, Some(port)),
            None => (authority, None),
        }
    };
    if host.is_empty() {
        return None;
    }
    let port = match port {
        None => None,
        Some(port) if !port.is_empty() && port.bytes().all(|b| b.is_ascii_digit()) => {
            Some(port.parse::<u16>().ok()?)
        }
        Some(_) => return None,
    };
    Some((host, port))
}

/// Whether `label` could be one part of a host name. Anything outside ASCII
/// is left for the URL parser to judge, so "münchen.de" is a site.
fn valid_label(label: &str) -> bool {
    !label.is_empty()
        && label
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_' || !c.is_ascii())
}

/// Whether `tld` is an ending the bar recognises: a listed one, or an
/// internationalised one, which are too many to list and never collide with
/// a file extension.
fn known_tld(tld: &str) -> bool {
    tld.starts_with("xn--")
        || !tld.is_ascii()
        || TLDS
            .lines()
            .filter(|line| !line.starts_with('#'))
            .flat_map(str::split_whitespace)
            .any(|known| known == tld)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;

    /// The examples the chrome's own copy of these rules is tested against.
    #[test]
    fn agrees_with_the_shared_examples() {
        let vectors: Vec<Value> =
            serde_json::from_str(include_str!("omnibox/vectors.json")).unwrap();
        assert!(vectors.len() > 40);
        for vector in vectors {
            let input = vector["input"].as_str().unwrap();
            if vector["unix"].as_bool() == Some(true) && cfg!(windows) {
                continue;
            }
            let got = classify(input);
            if vector["empty"].as_bool() == Some(true) {
                assert_eq!(got, None, "{input:?}");
            } else if let Some(search) = vector["search"].as_str() {
                assert_eq!(got, Some(Typed::Search(search.to_owned())), "{input:?}");
            } else if let Some(url) = vector["url"].as_str() {
                match got {
                    Some(Typed::Address(got)) => assert_eq!(got.as_str(), url, "{input:?}"),
                    other => panic!("{input:?} should open {url}, got {other:?}"),
                }
            } else if let Some(tail) = vector["file"].as_str() {
                match got {
                    Some(Typed::Address(got)) => {
                        assert_eq!(got.scheme(), "file", "{input:?}");
                        assert!(got.path().ends_with(tail), "{input:?} gave {got}");
                    }
                    other => panic!("{input:?} should open a file, got {other:?}"),
                }
            } else {
                panic!("vector {vector} says nothing about the outcome");
            }
        }
    }

    #[test]
    fn the_listed_endings_are_lowercase_and_unique() {
        let listed: Vec<&str> = TLDS
            .lines()
            .filter(|line| !line.starts_with('#'))
            .flat_map(str::split_whitespace)
            .collect();
        let mut unique = listed.clone();
        unique.sort_unstable();
        unique.dedup();
        assert_eq!(unique.len(), listed.len(), "a domain is listed twice");
        assert!(listed.iter().all(|tld| *tld == tld.to_ascii_lowercase()));
        assert!(!known_tld("js") && !known_tld("14") && known_tld("uk"));
    }

    #[test]
    fn a_windows_drive_is_never_taken_for_another_app() {
        // "C:" parses as a scheme; it must not become an app link.
        assert!(!matches!(
            classify(r"C:\Users\me"),
            Some(Typed::Address(url)) if url.scheme() == "c"
        ));
    }
}
